### PR TITLE
More benchmarks

### PR DESCRIPTION
**Pull request by pepeiborra**
This extends the benchmark suite with 
- two new experiments
- fixes for some of the issues reported in #617 
- Basic memory usage stats (unreliable due to https://github.com/bubba/lsp-test/issues/70)
- much better ouptut
- additional config flags, which I'm using to automate perf analysis elsewhere

### CODE REVIEW

File: ghcide.cabal [hunk 1]
```diff
@@ -319,7 +319,7 @@ benchmark ghcide-bench
         extra,
         filepath,
         ghcide,
-        lsp-test < 0.12,
+        lsp-test <= 0.11.0.2 && < 0.12,
```
Comment (cocreature) [inline ghcide.cabal:322]: That looks wrong, did you mean `>= 0.11.0.2`?

File: bench/Main.hs [hunk 2]
```diff
@@ -274,14 +372,37 @@ setup = do
...
 
   return $ removeDirectoryRecursive examplesPath
 
--- | Asks the server to shutdown and exit politely
-exitServer :: Session ()
-exitServer = request_ Shutdown (Nothing :: Maybe Value) >> sendNotification Exit ExitParams
-
 --------------------------------------------------------------------------------------------
+
+-- Parse the max residency in RTS -s output
```
Comment (cocreature) [inline bench/Main.hs:383]: Totally fine to keep this for now but once this starts to become more complex we should switch to the machine-readable output of the RTS stats.

File: ghcide.cabal (hunk 1 above)
Comment (pepeiborra) [reply ghcide.cabal:322]: yes, thanks
```suggestion
        lsp-test >= 0.11.0.2 && < 0.12,
```

File: bench/Main.hs (hunk 2 above)
Comment (pepeiborra) [reply bench/Main.hs:383]: Do you mean eventlog? It requires relinking the binary against the eventlog RTS, which would break `cabal bench`.

Comment (cocreature) [reply bench/Main.hs:383]: No I mean the machine readable output you get from the standard RTS, see https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/runtime_control.html#rts-options-to-produce-runtime-statistics

Comment (pepeiborra) [reply bench/Main.hs:383]: Oh, that's a good point, thanks for the tip! Definitely worth switching to that now or at some point.

**Pull request merged**